### PR TITLE
fix: downsample, scale and rotate all images before sending according to intended use - AR-1713 AR-1806

### PR DESCRIPTION
**Pull request by tmspzz**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-1713" title="AR-1713" target="_blank">BugAR-1713</a>  Image orientation when sending image is not respected (EXIF data)
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Images were sent uncompressed and unscaled resulting in degraded performance on all clients

### Solutions

Before sending an image, or uploading an avatar, [downsample it and reduce quality with logic similar to iOS](https://github.com/wireapp/wire-ios-images/blob/develop/Sources/Image%20Processing/ZMImageDownsampleOperation.m) 

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-689/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-689/2/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3685627)(3685627)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3685627)(3685627)-dev-release.apk)

**Review by gongracr — APPROVED** [review]
Looking real good 👏 ! Just some comments about naming and class type.
But nothing major so preapproving too! 🍉 🌞

**Review by vitorhugods — COMMENTED** [review]
I just wonder the same about the `sealed class` thingy. And it would be good to clarify the impacts of having width/heigth = 0.

Other than that, _che bello_!

File: app/src/main/kotlin/com/wire/android/util/ImageUtil.kt [hunk 1]
```diff
@@ -5,24 +5,97 @@ import android.content.ContentResolver.SCHEME_FILE
...
+                val output = ByteArrayOutputStream()
+                resizedImage.compress(Bitmap.CompressFormat.JPEG, qualityForSizeClass(sizeClass), output)
+                return output.toByteArray()
+            }
+            return byteArray
+        }
+
+        @Suppress("MagicNumber")
+        private fun scaledSizeForBitmap(bitmap: Bitmap, targetDimension: Float): Pair<Float, Float> {
+            val scale1 = java.lang.Float.max(targetDimension / bitmap.width, targetDimension / bitmap.height)
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/util/ImageUtil.kt]: I think we are able to use `kotlin.math.max`, is in the kotlin lib =D

**Review by yamilmedina — APPROVED** [review]
Looks amazing! Just small comments, great work on the (annoying) exif handling 👏  🥇

File: app/src/main/kotlin/com/wire/android/util/ImageUtil.kt [hunk 2]
```diff
@@ -5,24 +5,97 @@ import android.content.ContentResolver.SCHEME_FILE
...
+            val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
+            val exifInterface = ExifInterface(byteArray.inputStream())
+            val targetDimension = dimensionForSizeClass(sizeClass)
+            if (shouldScale(bitmap, targetDimension)) {
+                val size = scaledSizeForBitmap(bitmap, targetDimension)
+                val resizedImage = Bitmap
+                    .createScaledBitmap(bitmap, size.first.toInt(), size.second.toInt(), true)
+                    .rotateImageToNormalOrientation(exifInterface)
+                val output = ByteArrayOutputStream()
+                resizedImage.compress(Bitmap.CompressFormat.JPEG, qualityForSizeClass(sizeClass), output)
```
[the remark this answers is not in the document]
Comment (tmspzz) [reply app/src/main/kotlin/com/wire/android/util/ImageUtil.kt]: if image.hasAlpha() -> PNG

File: app/src/main/kotlin/com/wire/android/util/ImageUtil.kt [hunk 3]
```diff
@@ -5,24 +5,97 @@ import android.content.ContentResolver.SCHEME_FILE
...
+            return when (sizeClass) {
+                ImageSizeClass.Small -> 75
+                ImageSizeClass.Medium -> 45
+            }
+        }
+
+        fun extractImageWidthAndHeight(imageRawData: ByteArray): Pair<Int, Int> {
+            val exifInterface = ExifInterface(imageRawData.inputStream())
+            val exifWidth: Int = exifInterface.getAttributeInt(ExifInterface.TAG_IMAGE_WIDTH, 0)
+            val exifHeight: Int = exifInterface.getAttributeInt(ExifInterface.TAG_IMAGE_LENGTH, 0)
```
[the remark this answers is not in the document]
Comment (tmspzz) [reply app/src/main/kotlin/com/wire/android/util/ImageUtil.kt]: It seems to be able to make out the image size even for screenshot (for example). The size is only being used by the receiver to allocate screen real estate in advance

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-689/5/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv0.0.1(3685876)(3685876)-dev-release.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/dev/release/com.wire.android-vv0.0.1(3685876)(3685876)-dev-release.apk)

**Pull request merged**